File: 0440/solution.java
class Solution {
    public int findKthNumber(int n, int k) {
        k--;
        long curr = 1;
        while (k > 0) {
            long count = countNodes(curr, n);
            if (k >= count) {
                curr++;
                k -= count;
            } else {
                curr *= 10;
                k--;
            }
        }
        return (int) curr;
    }

    private long countNodes(long curr, int n) {
        long nextNode = curr + 1;
        long result = 0;
        while (curr <= n) {
            result += Math.min(n - curr + 1, nextNode - curr);
            curr *= 10;
            nextNode *= 10;
        }
        return result;
    }
}

// time O(log^2(n))
// space O(1)

File: 0440/solution.py
class Solution:
    def findKthNumber(self, n: int, k: int) -> int:
        k -= 1
        curr = 1

        def subtreeNodes(num):
            res = 0
            nextNum = num + 1
            while num <= n:
                res += min(n - num + 1, nextNum - num)
                num *= 10
                nextNum *= 10
            return res

        while k:
            count = subtreeNodes(curr)
            if k >= count:
                curr += 1
                k -= count
            else:
                curr *= 10
                k -= 1
        return curr
